### PR TITLE
Modification for adoption by PyTorch

### PR DESCRIPTION
**Pull request by shibatch**
* This patch adds the following CMAKE variables
  - BUILD_GNUABI_LIBS : Turned on by default.
  - BUILD_TESTS : Turned on by default.
  - ~~BUILD_POSITION_INDEPENDENT_CODE : Turned on by default.~~ Position independent code is generated only if shared libs are built.
  - ~~SLEEF_DISABLE_AVX512F : Turned off by default.~~ The bug in gcc-5 with _mm512_extracti64x4_epi64 will be automatically detected.
* Travis build should now succeed if environment variable SDE_URL is not defined.
* Intrinsic functions that requires constant immediate value as an argument are now defined macros in helper files.
* fabs functions in sleefdp.c are now replaced with fabsk.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 option(BUILD_SHARED_LIBS "Build shared libs" ON)
 option(BUILD_DFT "libsleefdft will be built." ON)
+option(BUILD_GNUABI_LIBS "libsleefgnuabi will be built." ON)
+option(BUILD_TESTS "Tests will be built." ON)
+option(BUILD_POSITION_INDEPENDENT_CODE "Generates position independent code." ON)
```
[the remark this answers is not in the document]
Comment (shibatch) [reply CMakeLists.txt]: I removed BUILD_POSITION_INDEPENDENT_CODE. -fPIC is now specified only when shared libs are built. Do you think we need a static lib builder?

File: src/common/CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,13 @@
 # Compiler properties
 
 set(COMMON_TARGET_PROPERTIES 
-  POSITION_INDEPENDENT_CODE ON   # -fPIC
   C_STANDARD 99                  # -std=gnu99
   )
 
+if (BUILD_SHARED_LIBS)
+  list(APPEND COMMON_TARGET_PROPERTIES POSITION_INDEPENDENT_CODE ON)   # -fPIC
```
Comment (fpetrogalli-arm) [inline src/common/CMakeLists.txt]: I believe that `BUILD_SHARED_LIBS` already implies `-fPIC` on targets that support it. I think we should remove this, also in all other places.

**Comment by cpuhrsch** [comment]
This PR appears to fix every [on our end](https://github.com/pytorch/pytorch/pull/7341) except for [one error](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-cuda9-cudnn7-py3-build/8273/console):

```
18:40:11 cd /var/lib/jenkins/cpp-build/aten/sleef/src/libm && /opt/cache/bin/cc -DDORENAME=1 -DENABLE_ALIAS=1 -DENABLE_AVX2128=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMAGMA_V2 -DSLEEF_STATIC_LIBS=1 -DTH_BLAS_MKL -DTH_INDEX_BASE=0 -DUSE_GCC_ATOMICS=1 -D_FILE_OFFSET_BITS=64 -I/opt/conda/include -I/var/lib/jenkins/workspace/aten/src/TH -I/var/lib/jenkins/workspace/aten/src/THC -I/var/lib/jenkins/cpp-build/aten/src/TH -I/var/lib/jenkins/cpp-build/aten/src/THC -I/var/lib/jenkins/workspace/aten/src/THS -I/var/lib/jenkins/workspace/aten/src/THCS -I/var/lib/jenkins/cpp-build/aten/src/THS -I/var/lib/jenkins/cpp-build/aten/src/THCS -I/var/lib/jenkins/workspace/aten/src -I/var/lib/jenkins/cpp-build/aten/src -I/usr/local/cuda/include -I/var/lib/jenkins/workspace/aten/src/THNN -I/var/lib/jenkins/workspace/aten/src/THCUNN -I/TH -I/var/lib/jenkins/workspace/aten/src/ATen/cuda -I/var/lib/jenkins/workspace/aten/src/ATen/.. -I/var/lib/jenkins/cpp-build/aten/src/ATen -I/var/lib/jenkins/workspace/third_party/sleef/src/common -I/var/lib/jenkins/workspace/third_party/sleef/src/arch -I/var/lib/jenkins/cpp-build/aten/sleef/src/libm/include -I/var/lib/jenkins/workspace/third_party/sleef/src/libm  -Wall -Wno-unused -Wno-attributes -Wno-unused-result -Wno-psabi -ffp-contract=off -fno-math-errno -fno-trapping-math -g -fPIC   -DCUDA_HAS_FP16=1 -DHAVE_GCC_GET_CPUID -DUSE_SSE2 -DUSE_SSE4_1 -DUSE_SSE4_2 -DUSE_SSE3 -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -mavx2 -mfma -std=gnu99 -o CMakeFiles/sleefavx2128.dir/sleefsimdsp.c.o   -c /var/lib/jenkins/workspace/third_party/sleef/src/libm/sleefsimdsp.c
18:40:11 /var/lib/jenkins/workspace/third_party/sleef/src/libm/sleefsimddp.c: In function 'Sleef_sind8_u35avx512f':
18:40:11 /var/lib/jenkins/workspace/third_party/sleef/src/libm/sleefsimddp.c:294:1: error: unrecognizable insn:
18:40:11  }
18:40:11  ^
18:40:11 (insn 43 42 44 2 (set (mem/c:V4DF (plus:DI (reg/f:DI 82 virtual-stack-vars)
18:40:11                 (const_int -13504 [0xffffffffffffcb40])) [0 hi+0 S32 A256])
18:40:11         (vec_merge:V4DF (vec_select:V4DF (reg:V8DF 282)
18:40:11                 (parallel [
18:40:11                         (const_int 4 [0x4])
18:40:11                         (const_int 5 [0x5])
18:40:11                         (const_int 6 [0x6])
18:40:11                         (const_int 7 [0x7])
18:40:11                     ]))
18:40:11             (reg:V4DF 263 [ D.156477 ])
18:40:11             (reg:QI 283))) /var/lib/jenkins/workspace/third_party/sleef/src/arch/helperavx512f.h:123 -1
18:40:11      (nil))
18:40:11 /var/lib/jenkins/workspace/third_party/sleef/src/libm/sleefsimddp.c:294:1: internal compiler error: in extract_insn, at recog.c:2343
18:40:11 Please submit a full bug report,
18:40:11 with preprocessed source if appropriate.
18:40:11 See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
18:40:11 sleef/src/libm/CMakeFiles/sleefavx512f.dir/build.make:91: recipe for target 'sleef/src/libm/CMakeFiles/sleefavx512f.dir/sleefsimddp.c.o' failed
18:40:11 make[2]: Leaving directory '/var/lib/jenkins/cpp-build/aten'
18:40:11 make[2]: *** [sleef/src/libm/CMakeFiles/sleefavx512f.dir/sleefsimddp.c.o] Error 1
18:40:11 CMakeFiles/Makefile2:1148: recipe for target 'sleef/src/libm/CMakeFiles/sleefavx512f.dir/all' failed
18:40:11 make[1]: *** [sleef/src/libm/CMakeFiles/sleefavx512f.dir/all] Error 2
18:40:11 make[1]: *** Waiting for unfinished jobs....
```

File: src/common/CMakeLists.txt (hunk 2 above)
Comment (shibatch) [reply src/common/CMakeLists.txt]: It seems that's not the case.

**Comment by shibatch** [comment]
@cpuhrsch I added SLEEF_DISABLE_AVX512F cmake variable. Please check if this works.

File: src/common/CMakeLists.txt (hunk 2 above)
Comment (fpetrogalli-arm) [reply src/common/CMakeLists.txt]: I think -fPIC is added automatically. This is the behavior on my mac:
```
C02TC1ABHF1T:test frapet01$ cat lib.c
double f(double x) {
  return x + x * x;
}
C02TC1ABHF1T:test frapet01$ cat CMakeLists.txt
project(LIB LANGUAGES C)
add_library(mylib lib.c)
```
I build without `BUILD_SHARED_LIBS`, and I don't see `-fPIC`:
```
C02TC1ABHF1T:test frapet01$ mkdir build
C02TC1ABHF1T:test frapet01$ cd build/
C02TC1ABHF1T:build frapet01$ cmake ../
-- The C compiler identification is AppleClang 9.0.0.9000039
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
//...
-- Build files have been written to: /Users/frapet01/projects/test/build
C02TC1ABHF1T:build frapet01$ make VERBOSE=1
/...
[ 50%] Building C object CMakeFiles/mylib.dir/lib.c.o
/Library/Developer/CommandLineTools/usr/bin/cc    -o CMakeFiles/mylib.dir/lib.c.o   -c /Users/frapet01/projects/test/lib.c
[100%] Linking C static library libmylib.a
//...
C02TC1ABHF1T:build frapet01$
```

Then I build with `BUILD_SHARED_LIBS`, and as you can see, `-fPIC` is added:
```
C02TC1ABHF1T:test frapet01$ mkdir build-shared
C02TC1ABHF1T:test frapet01$ cd build-shared/
C02TC1ABHF1T:build-shared frapet01$ cmake ../ -DBUILD_SHARED_LIBS=On
-- The C compiler identification is AppleClang 9.0.0.9000039
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
// ...
-- Generating done
-- Build files have been written to: /Users/frapet01/projects/test/build-shared
C02TC1ABHF1T:build-shared frapet01$ make VERBOSE=1
// ...
[ 50%] Building C object CMakeFiles/mylib.dir/lib.c.o
/Library/Developer/CommandLineTools/usr/bin/cc -Dmylib_EXPORTS  -fPIC   -o CMakeFiles/mylib.dir/lib.c.o   -c /Users/frapet01/projects/test/lib.c
[100%] Linking C shared library libmylib.dylib
```

I think it makes sense, as `-fPIC` is described as the flag needed to build libraries, so cmake is making a good job in adding it automatically when `BUILD_SHARED_LIBS=On`.

**Comment by fpetrogalli-arm** [comment]
@cpuhrsch @shibatch The failure reported in https://github.com/shibatch/sleef/pull/190#issuecomment-387189497 seems to be a compiler problem.

@cpuhrsch Which compiler are you using? If you are using a fairly old version of your favorite compiler, it might be worth trying a new version.

@shibatch I am not really keen in adding a custom cmake variable to turn off the AVX512f. I'd rather relay in the COMPILER_SUPPORT_AVX512F variable, making sure that the test in the `check_c_compiles` invocation captures the failure that @cpuhrsch is seeing.

**Comment by shibatch** [comment]
>@shibatch I am not really keen in adding a custom cmake variable to turn off the AVX512f. I'd rather relay in the COMPILER_SUPPORT_AVX512F variable, making sure that the test in the check_c_compiles invocation captures the failure that @cpuhrsch is seeing.

It's not very easy to figure out the condition under which an internal compiler error occurs. My plan is to do it later. Then, SLEEF_DISABLE_AVX512F variable can be safely removed.

**Comment by fpetrogalli-arm** [comment]
> It's not very easy to figure out the condition under which an internal compiler error occurs

I see. In fact I was hoping to get a reproducer from the error reported at line `src/libm/sleefsimddp.c:294`, but the error is placed _after_ the return statement, so you are right.

It would be good to know from @cpuhrsch whether they need to use this specific compiler, to see if they can update it, but given that the error is reported from a jenkins run, I suspect it is not easy for them to upgrade compiler...

So, yes,  I am happy for you to add `SLEEF_DISABLE_AVX512F`. :)

File: src/common/CMakeLists.txt (hunk 2 above)
Comment (shibatch) [reply src/common/CMakeLists.txt]: The current CMakeLists.txt first compiles the source files into object files, and then add them to a shared lib. So, it's like
```
project(LIB LANGUAGES C)
add_library(libobj OBJECT lib.c)
add_library(mylib $<TARGET_OBJECTS:libobj>)
```
In this case, -fPIC is not specified when the source file is compiled.

**Comment by shibatch** [comment]
It's this bug : https://www.mail-archive.com/gcc-bugs@gcc.gnu.org/msg550808.html
_mm512_extracti64x4_epi64 is used in xsin.

**Comment by cpuhrsch** [comment]
[All checks ran clear](https://github.com/pytorch/pytorch/pull/7341) on https://github.com/shibatch/sleef/tree/b0b497af3342cee7ef9746c3a5b5e87b2116b6ea

I'm trying the most recent commit now.

**Pull request merged**